File: app/src/main/java/org/schabi/newpipe/settings/tabs/Tab.java
package org.schabi.newpipe.settings.tabs;

import android.content.Context;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.grack.nanojson.JsonObject;
import com.grack.nanojson.JsonStringWriter;

import org.schabi.newpipe.R;
import org.schabi.newpipe.database.LocalItem.LocalItemType;
import org.schabi.newpipe.error.ErrorInfo;
import org.schabi.newpipe.error.ErrorUtil;
import org.schabi.newpipe.error.UserAction;
import org.schabi.newpipe.extractor.NewPipe;
import org.schabi.newpipe.extractor.StreamingService;
import org.schabi.newpipe.extractor.exceptions.ExtractionException;
import org.schabi.newpipe.fragments.BlankFragment;
import org.schabi.newpipe.fragments.list.channel.ChannelFragment;
import org.schabi.newpipe.fragments.list.kiosk.DefaultKioskFragment;
import org.schabi.newpipe.fragments.list.kiosk.KioskFragment;
import org.schabi.newpipe.fragments.list.playlist.PlaylistFragment;
import org.schabi.newpipe.local.bookmark.BookmarkFragment;
import org.schabi.newpipe.local.feed.FeedFragment;
import org.schabi.newpipe.local.history.StatisticsPlaylistFragment;
import org.schabi.newpipe.local.playlist.LocalPlaylistFragment;
import org.schabi.newpipe.local.subscription.SubscriptionFragment;
import org.schabi.newpipe.util.KioskTranslator;
import org.schabi.newpipe.util.ServiceHelper;

import java.util.Objects;

public abstract class Tab {
    private static final String JSON_TAB_ID_KEY = "tab_id";

    private static final String NO_NAME = "<no-name>";
    private static final String NO_ID = "<no-id>";
    private static final String NO_URL = "<no-url>";

    Tab() {
    }

    Tab(@NonNull final JsonObject jsonObject) {
        readDataFromJson(jsonObject);
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Tab Handling
    //////////////////////////////////////////////////////////////////////////*/

    @Nullable
    public static Tab from(@NonNull final JsonObject jsonObject) {
        final int tabId = jsonObject.getInt(Tab.JSON_TAB_ID_KEY, -1);

        if (tabId == -1) {
            return null;
        }

        return from(tabId, jsonObject);
    }

    @Nullable
    public static Tab from(final int tabId) {
        return from(tabId, null);
    }

    @Nullable
    public static Type typeFrom(final int tabId) {
        for (final Type available : Type.values()) {
            if (available.getTabId() == tabId) {
                return available;
            }
        }
        return null;
    }

    @Nullable
    private static Tab from(final int tabId, @Nullable final JsonObject jsonObject) {
        final Type type = typeFrom(tabId);

        if (type == null) {
            return null;
        }

        if (jsonObject != null) {
            switch (type) {
                case KIOSK:
                    return new KioskTab(jsonObject);
                case CHANNEL:
                    return new ChannelTab(jsonObject);
                case PLAYLIST:
                    return new PlaylistTab(jsonObject);
            }
        }

        return type.getTab();
    }

    public abstract int getTabId();

    public abstract String getTabName(Context context);

    @DrawableRes
    public abstract int getTabIconRes(Context context);

    /**
     * Return a instance of the fragment that this tab represent.
     *
     * @param context Android app context
     * @return the fragment this tab represents
     */
    public abstract Fragment getFragment(Context context) throws ExtractionException;

    @Override
    public boolean equals(final Object obj) {
        if (!(obj instanceof Tab)) {
            return false;
        }
        final Tab other = (Tab) obj;
        return getTabId() == other.getTabId();
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(getTabId());
    }

    /*//////////////////////////////////////////////////////////////////////////
    // JSON Handling
    //////////////////////////////////////////////////////////////////////////*/

    public void writeJsonOn(final JsonStringWriter jsonSink) {
        jsonSink.object();

        jsonSink.value(JSON_TAB_ID_KEY, getTabId());
        writeDataToJson(jsonSink);

        jsonSink.end();
    }

    protected void writeDataToJson(final JsonStringWriter writerSink) {
        // No-op
    }

    protected void readDataFromJson(final JsonObject jsonObject) {
        // No-op
    }

    /*//////////////////////////////////////////////////////////////////////////
    // Implementations
    //////////////////////////////////////////////////////////////////////////*/

    public enum Type {
        BLANK(new BlankTab()),
        DEFAULT_KIOSK(new DefaultKioskTab()),
        SUBSCRIPTIONS(new SubscriptionsTab()),
        FEED(new FeedTab()),
        BOOKMARKS(new BookmarksTab()),
        HISTORY(new HistoryTab()),
        KIOSK(new KioskTab()),
        CHANNEL(new ChannelTab()),
        PLAYLIST(new PlaylistTab());

        private final Tab tab;

        Type(final Tab tab) {
            this.tab = tab;
        }

        public int getTabId() {
            return tab.getTabId();
        }

        public Tab getTab() {
            return tab;
        }
    }

    public static class BlankTab extends Tab {
        public static final int ID = 0;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            // TODO: find a better name for the blank tab (maybe "blank_tab") or replace it with
            //       context.getString(R.string.app_name);
            return "NewPipe"; // context.getString(R.string.blank_page_summary);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_crop_portrait;
        }

        @Override
        public BlankFragment getFragment(final Context context) {
            return new BlankFragment();
        }
    }

    public static class SubscriptionsTab extends Tab {
        public static final int ID = 1;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return context.getString(R.string.tab_subscriptions);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_tv;
        }

        @Override
        public SubscriptionFragment getFragment(final Context context) {
            return new SubscriptionFragment();
        }

    }

    public static class FeedTab extends Tab {
        public static final int ID = 2;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return context.getString(R.string.fragment_feed_title);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_subscriptions;
        }

        @Override
        public FeedFragment getFragment(final Context context) {
            return new FeedFragment();
        }
    }

    public static class BookmarksTab extends Tab {
        public static final int ID = 3;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return context.getString(R.string.tab_bookmarks);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_bookmark;
        }

        @Override
        public BookmarkFragment getFragment(final Context context) {
            return new BookmarkFragment();
        }
    }

    public static class HistoryTab extends Tab {
        public static final int ID = 4;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return context.getString(R.string.title_activity_history);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_history;
        }

        @Override
        public StatisticsPlaylistFragment getFragment(final Context context) {
            return new StatisticsPlaylistFragment();
        }
    }

    public static class KioskTab extends Tab {
        public static final int ID = 5;
        private static final String JSON_KIOSK_SERVICE_ID_KEY = "service_id";
        private static final String JSON_KIOSK_ID_KEY = "kiosk_id";
        private int kioskServiceId;
        private String kioskId;

        private KioskTab() {
            this(-1, NO_ID);
        }

        public KioskTab(final int kioskServiceId, final String kioskId) {
            this.kioskServiceId = kioskServiceId;
            this.kioskId = kioskId;
        }

        public KioskTab(final JsonObject jsonObject) {
            super(jsonObject);
        }

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return KioskTranslator.getTranslatedKioskName(kioskId, context);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            final int kioskIcon = KioskTranslator.getKioskIcon(kioskId);

            if (kioskIcon <= 0) {
                throw new IllegalStateException("Kiosk ID is not valid: \"" + kioskId + "\"");
            }

            return kioskIcon;
        }

        @Override
        public KioskFragment getFragment(final Context context) throws ExtractionException {
            return KioskFragment.getInstance(kioskServiceId, kioskId);
        }

        @Override
        protected void writeDataToJson(final JsonStringWriter writerSink) {
            writerSink.value(JSON_KIOSK_SERVICE_ID_KEY, kioskServiceId)
                    .value(JSON_KIOSK_ID_KEY, kioskId);
        }

        @Override
        protected void readDataFromJson(final JsonObject jsonObject) {
            kioskServiceId = jsonObject.getInt(JSON_KIOSK_SERVICE_ID_KEY, -1);
            kioskId = jsonObject.getString(JSON_KIOSK_ID_KEY, NO_ID);
        }

        @Override
        public boolean equals(final Object obj) {
            if (!(obj instanceof KioskTab)) {
                return false;
            }
            final KioskTab other = (KioskTab) obj;
            return super.equals(obj)
                    && kioskServiceId == other.kioskServiceId
                    && kioskId.equals(other.kioskId);
        }

        @Override
        public int hashCode() {
            return Objects.hash(getTabId(), kioskServiceId, kioskId);
        }

        public int getKioskServiceId() {
            return kioskServiceId;
        }

        public String getKioskId() {
            return kioskId;
        }
    }

    public static class ChannelTab extends Tab {
        public static final int ID = 6;
        private static final String JSON_CHANNEL_SERVICE_ID_KEY = "channel_service_id";
        private static final String JSON_CHANNEL_URL_KEY = "channel_url";
        private static final String JSON_CHANNEL_NAME_KEY = "channel_name";
        private int channelServiceId;
        private String channelUrl;
        private String channelName;

        private ChannelTab() {
            this(-1, NO_URL, NO_NAME);
        }

        public ChannelTab(final int channelServiceId, final String channelUrl,
                          final String channelName) {
            this.channelServiceId = channelServiceId;
            this.channelUrl = channelUrl;
            this.channelName = channelName;
        }

        public ChannelTab(final JsonObject jsonObject) {
            super(jsonObject);
        }

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return channelName;
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_tv;
        }

        @Override
        public ChannelFragment getFragment(final Context context) {
            return ChannelFragment.getInstance(channelServiceId, channelUrl, channelName);
        }

        @Override
        protected void writeDataToJson(final JsonStringWriter writerSink) {
            writerSink.value(JSON_CHANNEL_SERVICE_ID_KEY, channelServiceId)
                    .value(JSON_CHANNEL_URL_KEY, channelUrl)
                    .value(JSON_CHANNEL_NAME_KEY, channelName);
        }

        @Override
        protected void readDataFromJson(final JsonObject jsonObject) {
            channelServiceId = jsonObject.getInt(JSON_CHANNEL_SERVICE_ID_KEY, -1);
            channelUrl = jsonObject.getString(JSON_CHANNEL_URL_KEY, NO_URL);
            channelName = jsonObject.getString(JSON_CHANNEL_NAME_KEY, NO_NAME);
        }

        @Override
        public boolean equals(final Object obj) {
            if (!(obj instanceof ChannelTab)) {
                return false;
            }
            final ChannelTab other = (ChannelTab) obj;
            return super.equals(obj)
                    && channelServiceId == other.channelServiceId
                    && channelUrl.equals(other.channelName)
                    && channelName.equals(other.channelName);
        }

        @Override
        public int hashCode() {
            return Objects.hash(getTabId(), channelServiceId, channelUrl, channelName);
        }

        public int getChannelServiceId() {
            return channelServiceId;
        }

        public String getChannelUrl() {
            return channelUrl;
        }

        public String getChannelName() {
            return channelName;
        }
    }

    public static class DefaultKioskTab extends Tab {
        public static final int ID = 7;

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return KioskTranslator.getTranslatedKioskName(getDefaultKioskId(context), context);
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return KioskTranslator.getKioskIcon(getDefaultKioskId(context));
        }

        @Override
        public DefaultKioskFragment getFragment(final Context context) {
            return new DefaultKioskFragment();
        }

        private String getDefaultKioskId(final Context context) {
            final int kioskServiceId = ServiceHelper.getSelectedServiceId(context);

            String kioskId = "";
            try {
                final StreamingService service = NewPipe.getService(kioskServiceId);
                kioskId = service.getKioskList().getDefaultKioskId();
            } catch (final ExtractionException e) {
                ErrorUtil.showSnackbar(context, new ErrorInfo(e,
                        UserAction.REQUESTED_KIOSK, "Loading default kiosk for selected service"));
            }
            return kioskId;
        }
    }

    public static class PlaylistTab extends Tab {
        public static final int ID = 8;
        private static final String JSON_PLAYLIST_SERVICE_ID_KEY = "playlist_service_id";
        private static final String JSON_PLAYLIST_URL_KEY = "playlist_url";
        private static final String JSON_PLAYLIST_NAME_KEY = "playlist_name";
        private static final String JSON_PLAYLIST_ID_KEY = "playlist_id";
        private static final String JSON_PLAYLIST_TYPE_KEY = "playlist_type";
        private int playlistServiceId;
        private String playlistUrl;
        private String playlistName;
        private long playlistId;
        private LocalItemType playlistType;

        private PlaylistTab() {
            this(-1, NO_NAME);
        }

        public PlaylistTab(final long playlistId, final String playlistName) {
            this.playlistName = playlistName;
            this.playlistId = playlistId;
            this.playlistType = LocalItemType.PLAYLIST_LOCAL_ITEM;
            this.playlistServiceId = -1;
            this.playlistUrl = NO_URL;
        }

        public PlaylistTab(final int playlistServiceId, final String playlistUrl,
                           final String playlistName) {
            this.playlistServiceId = playlistServiceId;
            this.playlistUrl = playlistUrl;
            this.playlistName = playlistName;
            this.playlistType = LocalItemType.PLAYLIST_REMOTE_ITEM;
            this.playlistId = -1;
        }

        public PlaylistTab(final JsonObject jsonObject) {
            super(jsonObject);
        }

        @Override
        public int getTabId() {
            return ID;
        }

        @Override
        public String getTabName(final Context context) {
            return playlistName;
        }

        @DrawableRes
        @Override
        public int getTabIconRes(final Context context) {
            return R.drawable.ic_bookmark;
        }

        @Override
        public Fragment getFragment(final Context context) {
            if (playlistType == LocalItemType.PLAYLIST_LOCAL_ITEM) {
                return LocalPlaylistFragment.getInstance(playlistId, playlistName);

            } else { // playlistType == LocalItemType.PLAYLIST_REMOTE_ITEM
                return PlaylistFragment.getInstance(playlistServiceId, playlistUrl, playlistName);
            }
        }

        @Override
        protected void writeDataToJson(final JsonStringWriter writerSink) {
            writerSink.value(JSON_PLAYLIST_SERVICE_ID_KEY, playlistServiceId)
                    .value(JSON_PLAYLIST_URL_KEY, playlistUrl)
                    .value(JSON_PLAYLIST_NAME_KEY, playlistName)
                    .value(JSON_PLAYLIST_ID_KEY, playlistId)
                    .value(JSON_PLAYLIST_TYPE_KEY, playlistType.toString());
        }

        @Override
        protected void readDataFromJson(final JsonObject jsonObject) {
            playlistServiceId = jsonObject.getInt(JSON_PLAYLIST_SERVICE_ID_KEY, -1);
            playlistUrl = jsonObject.getString(JSON_PLAYLIST_URL_KEY, NO_URL);
            playlistName = jsonObject.getString(JSON_PLAYLIST_NAME_KEY, NO_NAME);
            playlistId = jsonObject.getInt(JSON_PLAYLIST_ID_KEY, -1);
            playlistType = LocalItemType.valueOf(
                    jsonObject.getString(JSON_PLAYLIST_TYPE_KEY,
                            LocalItemType.PLAYLIST_LOCAL_ITEM.toString())
            );
        }

        @Override
        public boolean equals(final Object obj) {
            if (!(obj instanceof PlaylistTab)) {
                return false;
            }

            final PlaylistTab other = (PlaylistTab) obj;

            return super.equals(obj)
                    && playlistServiceId == other.playlistServiceId // Remote
                    && playlistId == other.playlistId // Local
                    && playlistUrl.equals(other.playlistUrl)
                    && playlistName.equals(other.playlistName)
                    && playlistType == other.playlistType;
        }

        @Override
        public int hashCode() {
            return Objects.hash(
                    getTabId(),
                    playlistServiceId,
                    playlistId,
                    playlistUrl,
                    playlistName,
                    playlistType
            );
        }

        public int getPlaylistServiceId() {
            return playlistServiceId;
        }

        public String getPlaylistUrl() {
            return playlistUrl;
        }

        public String getPlaylistName() {
            return playlistName;
        }

        public long getPlaylistId() {
            return playlistId;
        }

        public LocalItemType getPlaylistType() {
            return playlistType;
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/settings/tabs/Tab.kt
package org.schabi.newpipe.settings.tabs

import android.content.Context
import androidx.annotation.DrawableRes
import kotlinx.serialization.json.*
import java.util.*
import org.schabi.newpipe.R
import org.schabi.newpipe.database.LocalItem.LocalItemType
import org.schabi.newpipe.extractor.NewPipe
import org.schabi.newpipe.extractor.exceptions.ExtractionException
import org.schabi.newpipe.util.KioskTranslator
import org.schabi.newpipe.util.ServiceHelper

abstract class Tab {

    constructor()

    constructor(jsonObject: JsonObject) {
        readDataFromJson(jsonObject)
    }

    abstract val tabId: Int

    abstract fun getTabName(context: Context): String

    @DrawableRes
    abstract fun getTabIconRes(context: Context): Int

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Tab) return false
        return tabId == other.tabId
    }

    override fun hashCode(): Int {
        return Objects.hashCode(tabId)
    }

    fun writeJsonObject(): JsonObject {
        return buildJsonObject {
            put(JSON_TAB_ID_KEY, tabId)
            writeDataToJson(this)
        }
    }

    protected open fun writeDataToJson(builder: JsonObjectBuilder) {
        // No-op
    }

    protected open fun readDataFromJson(jsonObject: JsonObject) {
        // No-op
    }

    enum class Type(val tab: Tab) {
        BLANK(BlankTab()),
        DEFAULT_KIOSK(DefaultKioskTab()),
        SUBSCRIPTIONS(SubscriptionsTab()),
        FEED(FeedTab()),
        BOOKMARKS(BookmarksTab()),
        HISTORY(HistoryTab()),
        KIOSK(KioskTab()),
        CHANNEL(ChannelTab()),
        PLAYLIST(PlaylistTab()),
        FEEDGROUP(FeedGroupTab());

        val tabId: Int
            get() = tab.tabId
    }

    companion object {
        private const val JSON_TAB_ID_KEY = "tab_id"
        private const val NO_NAME = "<no-name>"
        private const val NO_ID = "<no-id>"
        private const val NO_URL = "<no-url>"

        @JvmStatic
        fun from(jsonObject: JsonObject): Tab? {
            val tabId = jsonObject[JSON_TAB_ID_KEY]?.jsonPrimitive?.int ?: -1
            return if (tabId == -1) null else from(tabId, jsonObject)
        }

        @JvmStatic
        fun from(tabId: Int): Tab? {
            return from(tabId, null)
        }

        @JvmStatic
        fun typeFrom(tabId: Int): Type? {
            return Type.entries.find { it.tabId == tabId }
        }

        private fun from(tabId: Int, jsonObject: JsonObject?): Tab? {
            val type = typeFrom(tabId) ?: return null

            if (jsonObject != null) {
                return when (type) {
                    Type.KIOSK -> KioskTab(jsonObject)
                    Type.CHANNEL -> ChannelTab(jsonObject)
                    Type.PLAYLIST -> PlaylistTab(jsonObject)
                    Type.FEEDGROUP -> FeedGroupTab(jsonObject)
                    else -> type.tab
                }
            }

            return type.tab
        }
    }

    class BlankTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = "NewPipe"

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_crop_portrait

        companion object {
            const val ID = 0
        }
    }

    class SubscriptionsTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = context.getString(R.string.tab_subscriptions)

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_tv

        companion object {
            const val ID = 1
        }
    }

    class FeedTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = context.getString(R.string.feed_title)

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_subscriptions

        companion object {
            const val ID = 2
        }
    }

    class BookmarksTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = context.getString(R.string.tab_bookmarks)

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_bookmark

        companion object {
            const val ID = 3
        }
    }

    class HistoryTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = context.getString(R.string.title_activity_history)

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_history

        companion object {
            const val ID = 4
        }
    }

    class KioskTab : Tab {
        var kioskServiceId: Int = -1
            private set
        var kioskId: String = NO_ID
            private set

        constructor() : super()

        constructor(kioskServiceId: Int, kioskId: String) : super() {
            this.kioskServiceId = kioskServiceId
            this.kioskId = kioskId
        }

        constructor(jsonObject: JsonObject) : super(jsonObject)

        override val tabId: Int = ID

        override fun getTabName(context: Context): String = KioskTranslator.getTranslatedKioskName(kioskId, context)

        override fun getTabIconRes(context: Context): Int {
            val kioskIcon = KioskTranslator.getKioskIcon(kioskId)
            if (kioskIcon <= 0) {
                throw IllegalStateException("Kiosk ID is not valid: \"$kioskId\"")
            }
            return kioskIcon
        }

        override fun writeDataToJson(builder: JsonObjectBuilder) {
            builder.put(JSON_KIOSK_SERVICE_ID_KEY, kioskServiceId)
            builder.put(JSON_KIOSK_ID_KEY, kioskId)
        }

        override fun readDataFromJson(jsonObject: JsonObject) {
            kioskServiceId = jsonObject[JSON_KIOSK_SERVICE_ID_KEY]?.jsonPrimitive?.int ?: -1
            kioskId = jsonObject[JSON_KIOSK_ID_KEY]?.jsonPrimitive?.content ?: NO_ID
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is KioskTab) return false
            if (!super.equals(other)) return false
            return kioskServiceId == other.kioskServiceId && kioskId == other.kioskId
        }

        override fun hashCode(): Int {
            return Objects.hash(tabId, kioskServiceId, kioskId)
        }

        companion object {
            const val ID = 5
            private const val JSON_KIOSK_SERVICE_ID_KEY = "service_id"
            private const val JSON_KIOSK_ID_KEY = "kiosk_id"
        }
    }

    class ChannelTab : Tab {
        var channelServiceId: Int = -1
            private set
        var channelUrl: String = NO_URL
            private set
        var channelName: String = NO_NAME
            private set

        constructor() : super()

        constructor(channelServiceId: Int, channelUrl: String, channelName: String) : super() {
            this.channelServiceId = channelServiceId
            this.channelUrl = channelUrl
            this.channelName = channelName
        }

        constructor(jsonObject: JsonObject) : super(jsonObject)

        override val tabId: Int = ID

        override fun getTabName(context: Context): String = channelName

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_tv

        override fun writeDataToJson(builder: JsonObjectBuilder) {
            builder.put(JSON_CHANNEL_SERVICE_ID_KEY, channelServiceId)
            builder.put(JSON_CHANNEL_URL_KEY, channelUrl)
            builder.put(JSON_CHANNEL_NAME_KEY, channelName)
        }

        override fun readDataFromJson(jsonObject: JsonObject) {
            channelServiceId = jsonObject[JSON_CHANNEL_SERVICE_ID_KEY]?.jsonPrimitive?.int ?: -1
            channelUrl = jsonObject[JSON_CHANNEL_URL_KEY]?.jsonPrimitive?.content ?: NO_URL
            channelName = jsonObject[JSON_CHANNEL_NAME_KEY]?.jsonPrimitive?.content ?: NO_NAME
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ChannelTab) return false
            if (!super.equals(other)) return false
            return channelServiceId == other.channelServiceId &&
                channelUrl == other.channelUrl &&
                channelName == other.channelName
        }

        override fun hashCode(): Int {
            return Objects.hash(tabId, channelServiceId, channelUrl, channelName)
        }

        companion object {
            const val ID = 6
            private const val JSON_CHANNEL_SERVICE_ID_KEY = "channel_service_id"
            private const val JSON_CHANNEL_URL_KEY = "channel_url"
            private const val JSON_CHANNEL_NAME_KEY = "channel_name"
        }
    }

    class DefaultKioskTab : Tab() {
        override val tabId: Int = ID

        override fun getTabName(context: Context): String = KioskTranslator.getTranslatedKioskName(getDefaultKioskId(context), context)

        override fun getTabIconRes(context: Context): Int = KioskTranslator.getKioskIcon(getDefaultKioskId(context))

        private fun getDefaultKioskId(context: Context): String {
            val kioskServiceId = ServiceHelper.getSelectedServiceId(context)
            return try {
                val service = NewPipe.getService(kioskServiceId)
                service.kioskList.defaultKioskId
            } catch (e: ExtractionException) {
                ""
            }
        }

        companion object {
            const val ID = 7
        }
    }

    class PlaylistTab : Tab {
        var playlistServiceId: Int = -1
            private set
        var playlistUrl: String = NO_URL
            private set
        var playlistName: String = NO_NAME
            private set
        var playlistId: Long = -1
            private set
        var playlistType: LocalItemType = LocalItemType.PLAYLIST_LOCAL_ITEM
            private set

        constructor() : super()

        constructor(playlistId: Long, playlistName: String) : super() {
            this.playlistName = playlistName
            this.playlistId = playlistId
            this.playlistType = LocalItemType.PLAYLIST_LOCAL_ITEM
            this.playlistServiceId = -1
            this.playlistUrl = NO_URL
        }

        constructor(playlistServiceId: Int, playlistUrl: String, playlistName: String) : super() {
            this.playlistServiceId = playlistServiceId
            this.playlistUrl = playlistUrl
            this.playlistName = playlistName
            this.playlistType = LocalItemType.PLAYLIST_REMOTE_ITEM
            this.playlistId = -1
        }

        constructor(jsonObject: JsonObject) : super(jsonObject)

        override val tabId: Int = ID

        override fun getTabName(context: Context): String = playlistName

        override fun getTabIconRes(context: Context): Int = R.drawable.ic_bookmark

        override fun writeDataToJson(builder: JsonObjectBuilder) {
            builder.put(JSON_PLAYLIST_SERVICE_ID_KEY, playlistServiceId)
            builder.put(JSON_PLAYLIST_URL_KEY, playlistUrl)
            builder.put(JSON_PLAYLIST_NAME_KEY, playlistName)
            builder.put(JSON_PLAYLIST_ID_KEY, playlistId)
            builder.put(JSON_PLAYLIST_TYPE_KEY, playlistType.toString())
        }

        override fun readDataFromJson(jsonObject: JsonObject) {
            playlistServiceId = jsonObject[JSON_PLAYLIST_SERVICE_ID_KEY]?.jsonPrimitive?.int ?: -1
            playlistUrl = jsonObject[JSON_PLAYLIST_URL_KEY]?.jsonPrimitive?.content ?: NO_URL
            playlistName = jsonObject[JSON_PLAYLIST_NAME_KEY]?.jsonPrimitive?.content ?: NO_NAME
            playlistId = jsonObject[JSON_PLAYLIST_ID_KEY]?.jsonPrimitive?.long ?: -1L
            playlistType = LocalItemType.valueOf(
                jsonObject[JSON_PLAYLIST_TYPE_KEY]?.jsonPrimitive?.content ?: LocalItemType.PLAYLIST_LOCAL_ITEM.toString()
            )
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is PlaylistTab) return false
            if (!super.equals(other)) return false
            return playlistServiceId == other.playlistServiceId &&
                playlistId == other.playlistId &&
                playlistUrl == other.playlistUrl &&
                playlistName == other.playlistName &&
                playlistType == other.playlistType
        }

        override fun hashCode(): Int {
            return Objects.hash(
                tabId,
                playlistServiceId,
                playlistId,
                playlistUrl,
                playlistName,
                playlistType
            )
        }

        companion object {
            const val ID = 8
            private const val JSON_PLAYLIST_SERVICE_ID_KEY = "playlist_service_id"
            private const val JSON_PLAYLIST_URL_KEY = "playlist_url"
            private const val JSON_PLAYLIST_NAME_KEY = "playlist_name"
            private const val JSON_PLAYLIST_ID_KEY = "playlist_id"
            private const val JSON_PLAYLIST_TYPE_KEY = "playlist_type"
        }
    }

    class FeedGroupTab : Tab {
        var feedGroupId: Long? = -1L
            private set
        var feedGroupName: String = NO_NAME
            private set
        var iconId: Int = R.drawable.ic_asterisk
            private set

        constructor() : super()

        constructor(feedGroupId: Long?, feedGroupName: String, iconId: Int) : super() {
            this.feedGroupId = feedGroupId
            this.feedGroupName = feedGroupName
            this.iconId = iconId
        }

        constructor(jsonObject: JsonObject) : super(jsonObject)

        override val tabId: Int = ID

        override fun getTabName(context: Context): String = context.getString(R.string.feed_title)

        override fun getTabIconRes(context: Context): Int = iconId

        override fun writeDataToJson(builder: JsonObjectBuilder) {
            builder.put(JSON_FEED_GROUP_ID_KEY, feedGroupId)
            builder.put(JSON_FEED_GROUP_NAME_KEY, feedGroupName)
            builder.put(JSON_FEED_GROUP_ICON_KEY, iconId)
        }

        override fun readDataFromJson(jsonObject: JsonObject) {
            feedGroupId = jsonObject[JSON_FEED_GROUP_ID_KEY]?.jsonPrimitive?.longOrNull ?: -1L
            feedGroupName = jsonObject[JSON_FEED_GROUP_NAME_KEY]?.jsonPrimitive?.content ?: NO_NAME
            iconId = jsonObject[JSON_FEED_GROUP_ICON_KEY]?.jsonPrimitive?.int ?: R.drawable.ic_asterisk
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is FeedGroupTab) return false
            if (!super.equals(other)) return false
            return feedGroupId == other.feedGroupId &&
                feedGroupName == other.feedGroupName &&
                iconId == other.iconId
        }

        override fun hashCode(): Int {
            return Objects.hash(tabId, feedGroupId, feedGroupName, iconId)
        }

        companion object {
            const val ID = 9
            private const val JSON_FEED_GROUP_ID_KEY = "feed_group_id"
            private const val JSON_FEED_GROUP_NAME_KEY = "feed_group_name"
            private const val JSON_FEED_GROUP_ICON_KEY = "feed_group_icon"
        }
    }
}
